stock_archive_constraint/models/product_product.py

```diff
@@ -0,0 +1,44 @@
+# Copyright 2020 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    @api.constrains("active")
+    def _check_active_stock_archive_constraint(self):
+        res = self.env['stock.quant'].search(
+            [
+                ('product_id', 'in', self.ids),
+                ('quantity', '!=', 0.0)
+            ], limit=1
+        )
+        if res:
+            raise ValidationError(
+                _("It is not possible to archive a product that has quant associated.")
+            )
+        res = self.env['stock.move'].search(
+            [
+                ('product_id', 'in', self.ids),
+                ('state', 'not in', ('done', 'cancel'))
+            ], limit=1
+        )
+        if res:
+            raise ValidationError(
+                _("It is not possible to archive a location that has "
+                  "stock move associated.")
```

Comment:
Same logic than previous comment + this is not covered by tests. Why is this in the same constraint? Please split them for convenience of inheritance:
```suggestion
                _("It is not possible to archive location '%s' which has "
                   "associated stock moves.")
```